prow/tide/search.go

```diff
@@ -0,0 +1,147 @@
+			totalMatches = int(sq.Search.IssueCount)
+			// If the search won't return all results, abort.
+			if totalMatches > 1000 {
+				return nil, totalMatches, nil
+			}
+			for _, n := range sq.Search.Nodes {
+				ret = append(ret, n.PullRequest)
+			}
+			if !sq.Search.PageInfo.HasNextPage {
+				break
+			}
+			vars["searchCursor"] = githubql.NewString(sq.Search.PageInfo.EndCursor)
+		}
+		log.Debugf("Search for query \"%s\" returned %d PRs and cost %d point(s). %d remaining.", q, len(ret), totalCost, remaining)
+		return ret, totalMatches, nil
+	}
+}
+
+func (q searchExecutor) search() ([]PullRequest, error) {
+	prs, _, err := q.searchRange(time.Time{}, time.Now())
+	return prs, err
+}
+
+func (q searchExecutor) searchSince(t time.Time) ([]PullRequest, error) {
+	prs, _, err := q.searchRange(t, time.Now())
+	return prs, err
+}
+
+func (q searchExecutor) searchRange(start, end time.Time) ([]PullRequest, int, error) {
+	prs, count, err := q(start, end)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	if count <= 1000 {
+		// The search returned all the results for the query.
+		return prs, len(prs), nil
+	}
+	// The query returned too many results, we need to partition it.
+	prs, err = q.partitionSearchRange(start, end, count)
+	return prs, len(prs), err
+}
+
+func (q searchExecutor) partitionSearchRange(start, end time.Time, count int) ([]PullRequest, error) {
+	partition := partitionTime(start, end, count, 900)
+	// Search right side...
+	rPRs, rCount, err := q.searchRange(partition, end)
+	if err != nil {
+		return nil, err
+	}
+
+	// Search left side...
+	// For the left side we can deduce the count in advance.
+	lCount := count - rCount
+	// If the count is too large we can skip the initial search and go straight to
+	// partitioning to save an API token.
+	var lPRs []PullRequest
+	if lCount <= 1000 {
+		lPRs, _, err = q.searchRange(start, partition)
+	} else {
+		lPRs, err = q.partitionSearchRange(start, partition, lCount)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	return append(lPRs, rPRs...), nil
+}
+
+func partitionTime(start, end time.Time, count, goalSize int) time.Time {
+	// TODO(cjwagner): Special case old start times. (Nobody has PRs from before 2000)
+	duration := end.Sub(start)
+	if count < goalSize*2 {
+		// Choose the midpoint.
+		return start.Add(duration / 2)
+	}
+	// Choose the point that will make the partitionTime->end range contain goalSize
+	// many results assuming a uniform distribution over time.
+	// ->    end - (duration * goalSize / count)
+	return end.Add((-duration * time.Duration(goalSize)) / time.Duration(count))
+}
+
+// dateToken generates a GitHub search query token for the specified date range.
+// See: https://help.github.com/articles/understanding-the-search-syntax/#query-for-dates
+func dateToken(start, end time.Time) string {
+	// Github's GraphQL API silently fails if you provide it with an invalid time
+	// string.
+	// Dates before 1970 are considered invalid.
```

Comment:
This isn't documented AFAIK so I don't have anything to link to. This was determined through experimentation when GitHub started throwing errors for dates that match ISO 8601, but were before 1970, the explanation is that the API silently fails.

Reply:
OK, this is almost definitely because they use unix epoch time somewhere though, fwiw.

We should write a doc somewhere with limitations like this that we've encountered experimentally, it would be really helpful for future GitHub tooling I think.